cmd/gen-go-sample/main.go

```diff
@@ -0,0 +1,227 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bytes"
+	"flag"
+	"fmt"
+	"go/format"
+	"io/ioutil"
+	"log"
+	"os"
+	"sort"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/protobuf/protoc-gen-go/descriptor"
+	"github.com/googleapis/gapic-generator-go/internal/errors"
+	"github.com/googleapis/gapic-generator-go/internal/pbinfo"
+	"github.com/googleapis/gapic-generator-go/internal/printer"
+	yaml "gopkg.in/yaml.v2"
+)
+
+func main() {
+	descFname := flag.String("desc", "", "proto descriptor")
+	gapicFname := flag.String("gapic", "", "gapic config")
+	nofmt := flag.Bool("nofmt", false, "skip gofmt, useful for debugging code with syntax error")
+	flag.Parse()
+
+	var (
+		gen = generator{
+			imps: map[pbinfo.ImportSpec]bool{},
+		}
+
+		donec = make(chan struct{})
+	)
+
+	go func() {
+		f, err := os.Open(*gapicFname)
+		if err != nil {
+			log.Fatal(errors.E(err, "cannot read GAPIC config file"))
+		}
+		defer f.Close()
+
+		if err := yaml.NewDecoder(f).Decode(&gen.gapic); err != nil {
+			log.Fatal(errors.E(err, "error reading GAPIC config file"))
+		}
+		donec <- struct{}{}
+	}()
+	go func() {
+		descBytes, err := ioutil.ReadFile(*descFname)
+		if err != nil {
+			log.Fatal(errors.E(err, "cannot read proto descriptor file"))
+		}
+
+		if err := proto.Unmarshal(descBytes, &gen.desc); err != nil {
+			log.Fatal(errors.E(err, "error reading proto descriptor file"))
+		}
+
+		gen.dinfo = pbinfo.Of(gen.desc.GetFile())
+		donec <- struct{}{}
+	}()
+
+	<-donec
+	<-donec
+
+	for _, iface := range gen.gapic.Interfaces {
+		for _, meth := range iface.Methods {
+			valSets := map[string]GapicValueSet{}
+			for _, vs := range meth.SampleValueSets {
+				valSets[vs.ID] = vs
+			}
+
+			for _, sam := range meth.Samples.Standalone {
+				for _, vsID := range sam.ValueSets {
+					vs, ok := valSets[vsID]
+					if !ok {
+						log.Fatal(errors.E(nil, "value set not found: %q", vsID))
+					}
+
+					gen.reset()
+					if err := gen.genSample(iface.Name, meth.Name, vs); err != nil {
+						log.Fatal(errors.E(err, "value set: %q", vs))
+					}
+					if err := gen.commit(!*nofmt); err != nil {
+						log.Fatal(errors.E(err, "value set: %q", vs))
+					}
+				}
+			}
+		}
+	}
+}
+
+type generator struct {
+	desc  descriptor.FileDescriptorSet
+	dinfo pbinfo.Info
+	gapic GapicConfig
+
+	pt   printer.P
+	imps map[pbinfo.ImportSpec]bool
```

Comment:
Could you describe the semantics of this map? What does the bool indicate?

Reply:
This is just a set of imports. We idiomatically map the imports to bool, so that code like
```
if imports[x] {
  // ...
}
```
work. We never assign `false`.